docs/developers/ui-components.rst

```diff
@@ -0,0 +1,301 @@
+such a component.
+
+Make sure to read up on how other organizations use components, which design constraints
+they have taken into consideration and where the variation lies if different community
+implementations exist.
+
+If nothing exists that is readily usable, you can proceed with implementing it yourself,
+but you should still keep in mind the NL DS research so that future implementations can
+converge on a single approach.
+
+Building a component from scratch
+=================================
+
+Composition
+-----------
+
+Often you need to build a higher level component that's composed of other lower level
+components. That's normal - the obvious low-level components like buttons and links are
+seldomly used in isolation, and we encourage re-using these instead of reinventing the
+wheel.
+
+Consider the footer example:
+
+.. code-block:: html
+
+    <footer class="openforms-footer">
+      <a href="#" class="utrecht-link">Link 1</a>
+      <a href="#" class="utrecht-link">Link 2</a>
+      <a href="#" class="utrecht-link">Link 3</a>
+    </footer>
+
+This simple example captures the essence well: the appearance of an individual link
+should be consistent with other links on the page, so you re-use the ``utrecht-link``
+component.
+
+However, the footer gets an application-specific class name, so that the layout and
+appearance (such as flex display with left or center alignment, background color...)
+can be specified on that element.
+
+Markup
+------
+
+Use semantic HTML, and try to use only what's necessary. Avoid wrapper ``div`` and
+``span`` elements if you can - often with Flexbox and CSS Grid styling you can get quite
+far for layout management.
+
+If you're not yet familiar with the implicit accessible roles of elements, take some
+time to do the research. E.g. it may seem obvious to use a ``nav`` element, but if this
+results in multiple such elements on a page, additional work is needed to give them
+distinctive accessible labels.
+
+The question "How would this appear to a screenreader" should always be at the forefront
+when writing HTML.
+
+Block-Element-Modifier (BEM)
+----------------------------
+
+`BEM <https://getbem.com/>`_ is a methodology for reusable components.
+
+**Blocks**
+
+For the CSS and CSS class names, the BEM methodology is used. In a nutshell, this means
+the outer element gets a ``.my-block`` classname that encapsulates the component and
+is the outer CSS target selector.
+
+Example:
+
+.. code-block:: html
+
+    <footer class="openforms-footer">
+        ...
+    </footer>
+
+.. tip::
+
+    * You should try to limit the number of different blocks that exist in a codebase.
+    * Avoid using extremely specific blocks that cannot be reused in other places.
+
+**Elements**
+
+The element can contain zero or more child elements, which get a ``.my-block__element-1``
+or ``.my-block__element-2`` class name. They can exist anywhere in the HTML tree, they
+are not limited to direct children.
```

Comment:
Could you please explain a bit this one?

Reply:
Pointed out in the docs that there's nesting happening here, even though in the BEM they seem siblings.